ambry-store/src/main/java/com.github.ambry.store/HardDeleter.java

```diff
@@ -111,18 +117,27 @@ public void run() {
     try {
       while (running.get()) {
         try {
-          if (!hardDelete()) {
-            isCaughtUp = true;
-            synchronized (lock) {
+          reentrantLock.lock();
```

Comment:
the lock call should be before the `try`(https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/ReentrantLock.html).